### ISSUE TITLE
Application Dependencies in Restart

### ISSUE THREAD

**Issue by sunethwarna** [KratosMultiphysics/Kratos#7088]
Hi,

I have encountered following situation. If I use a process from different application via json settings (i.e. StatisticsApplication) which uses variables from that app to store some nodal non-historical values, and I do a restart afterwards it seg-faults (my main script is fluid_dynamics_analysis.py where it doesn't know about StatApp being imported from processes). If I import StatApp in my MainKratos.py it works since it imports necessary variables lists. The question is, how can we make this more generic so we can keep main scripts as it is, and when saving restart files we can save some information about apps being used so that we can load them in resarting?

This problem will also be there if someone uses elements and/or conditions from different apps in runtime, where app is not imported via main script.

1. The first suggestion would be to move variables in utility apps like statistics app to core, but it won't solve the core problem of not being able to use two or more apps with restart.
2. There needs to be a way to store applications used in model part for restart, and those should be exposed to python level so we can read them and load them after searilizer is created.

Are there any other elegant ways of doing this?

I tagged the people I know who would be interested in a solution for this.

**Comment by philbucher** [comment]
My suggestion is to add a special method to the serializer with which one can get the names of the Apps that were loaded when the file was written
=> then when loading that file one can check or import the Apps that were loaded at the time of saving

I think this is not too much effort and will solve this problem in a very general way (same problem happens for EVERYTHING that gets registered in the `KratosComponents` from Apps)

**Comment by RiccardoRossi** [comment]
well, maybe we could add a check on this in the kernel save and load, justo
to verify that apps are registered correctly

**Comment by philbucher** [comment]
> well, maybe we could add a check on this in the kernel save and load, justo to verify that apps are registered correctly

but the kernel is not saved usually (at least afaik)

**Comment by RiccardoRossi** [comment]
@philbucher you are right, and i don't think we can really do a proper save, exactly because of the dependency on multiple apps, however MAYBE we could do some sanity checking...

**Comment by roigcarlo** [comment]
We could wrap the importlib calls in python so everything imported is logged and use that when storing/restoring.

This should free us to take care about dependencies, etc...

**Comment by RiccardoRossi** [comment]
We @KratosMultiphysics/technical-committee are thinking of introducing a `Project` object to include this kind of information.

This object would have to be serialized and would eventually contain all info about the project. Upon saving this object shall save the application list currently loaded, and upon loading it should throw a warning if the currently loaded applications do not match such a list

**Comment by RiccardoRossi** [comment]
we have been discussing this briefly in the @KratosMultiphysics/technical-committee .
The orchestrator is taking care of these issues by using pickle. take a look into 

https://github.com/KratosMultiphysics/Kratos/blob/06eeea305b06c8df084b031ec6415c66e4ed43d4/kratos/python_scripts/project.py#L113

for the solution we used there.

One "but" is that this solution is valid for python-based workflows and will not work as-is for people doing c++ only workflows.
